src/utils/eis_utils/src/eis_utils.c

```diff
@@ -468,208 +550,241 @@ EISUtilityResult RequestConnectionStringFromEISWithExpiry(
+        {
+            connType = ADUC_ConnType_Module;
+        }
 
-        if (certResult != EISErr_Ok)
+        // Build request for the signature
+        if (connType == ADUC_ConnType_Device)
+        {
+            resourceUri = ADUC_StringFormat("%s/devices/%s", hubName, deviceId);
+        }
+        else if (connType == ADUC_ConnType_Module)
+        {
+            resourceUri = ADUC_StringFormat("%s/devices/%s/modules/%s", hubName, deviceId, moduleId);
+        }
+        else
         {
-            result.err = certResult;
-            result.service = EISService_CertService;
             goto done;
         }
 
-        certResponseJson = json_parse_string(certResponseStr);
+        if (resourceUri == NULL)
+        {
+            goto done;
+        }
+
+        const char* gatewayHostName = json_object_get_string(specJson, EIS_IDENTITY_RESP_GATEWAYHOSTNAME_FIELD);
+
+        const JSON_Object* authJson =
+            json_value_get_object(json_object_get_value(specJson, EIS_IDENTITY_RESP_AUTH_FIELD));
 
-        if (certResponseJson == NULL)
+        if (authJson == NULL)
         {
             result.err = EISErr_InvalidJsonRespErr;
-            result.service = EISService_CertService;
+            result.service = EISService_IdentityService;
             goto done;
         }
 
-        const JSON_Object* certResponseJsonObj = json_value_get_object(certResponseJson);
-
-        const char* certificateStr = json_object_get_string(certResponseJsonObj, EIS_CERT_RESP_PEM);
+        const char* authTypeStr = json_object_get_string(authJson, EIS_IDENTITY_RESP_AUTH_TYPE_FIELD);
 
-        if (certificateStr == NULL)
+        if (authTypeStr == NULL)
         {
             result.err = EISErr_InvalidJsonRespErr;
-            result.service = EISService_CertService;
+            result.service = EISService_IdentityService;
             goto done;
         }
 
-        if (mallocAndStrcpy_s(&certString, certificateStr) != 0)
+        const char* keyHandle = json_object_get_string(authJson, EIS_IDENTITY_RESP_AUTH_KEYHANDLE_FIELD);
+
+        if (keyHandle == NULL)
         {
-            result.err = EISErr_ContentAllocErr;
+            result.err = EISErr_InvalidJsonRespErr;
+            result.service = EISService_KeyService;
             goto done;
         }
 
-        result.err =
-            BuildSasCertConnectionString(hubName, deviceId, moduleId, connType, gatewayHostName, &connectionStr);
+        if (strcmp(authTypeStr, "sas") == 0)
+        {
+            authType = ADUC_AuthType_SASToken;
+
+            result = BuildSharedAccessSignature(
+                resourceUri, keyHandle, expirySecsSinceEpoch, timeoutMS, &sharedSignatureStr);
+
+            if (result.err != EISErr_Ok)
+            {
+                goto done;
+            }
+
+            result.err = BuildSasTokenConnectionString(
+                hubName, deviceId, moduleId, connType, sharedSignatureStr, gatewayHostName, &connectionStr);
```

Comment:
Where are you requesting the construction of the sharedSignature? I don't see a call anywhere here to the key service?

Reply:
not changing anything on this part of the function. I think git just messed everything up! :(